packages/std/src/imports.rs

```diff
@@ -192,9 +197,12 @@ impl Api for ExternalApi {
 
         let result = unsafe { secp256k1_verify(hash_send_ptr, sig_send_ptr, pubkey_send_ptr) };
         match result {
-            0 => false,
-            1 => true,
-            _ => unreachable!(),
+            0 => Ok(true),
+            1 => Ok(false),
+            error_code => Err(StdError::crypto_err(format!(
+                "secp256k1_verify: error {}",
+                error_code
+            ))),
```

Comment:
Right now, this is the point where the newly added error message gets useless for the contract developer because it cannot be handles in a structured way.

Reply:
It may not be useful in the sense that it loses structure, but it's useful, in that it still gives more information.

If I understand you correctly, you want to add `VerificationError` above, so that error messages don't lose structure?

Reply:
Right, `VerificationError` can be handled without string parsing. The string representation is then only generated in order to make it StdResult compatible.

Reply:
My point is: if the original version (without error codes) already provides this information to the blockchain (in an unstructured way, that is), why bother here? Let's just stick with the version without error codes (IMO). 

I guess I'm asking this because I don't know how exactly all this plays out in the "real" (blockchain) world. What I want to avoid is not unstructured error messages, but lack of (specific, i.e. informative) error messages (for the end user, that is) at all.

Reply:
I see thwo options here:
1. Do not bother: `do_secp256k1_verify` creates a VmError that causes the transaction to fail (not crash), which is logged to chain
2. Return errors to the contract, allowing the contract developer handle or return as StdError

If we do 1. (as before), there is no point in returning the error cases to the contract (code > 1). If we do 2., we do all the work of creating the error reporting infrastructure. If we do all that, we should also allow the contract developer to handle the errors in a structured way.

Reply:
Yes, I totally agree that if doing 2., it makes sense to provide structured error handling.

What I'm proposing is doing 1. instead (that is, NOT doing 2.). Sorry for the detour, but it was because of a misunderstanding on my part. If the error is logged somewhere, and it can be recovered in "full form", so to speak, that's enough for me.

Reply:
I'm glad we're having this discussion. It is an important one and a hard one. This cannot easily be copied from existing code because we are introducing a new type of API calls that do not go contract -> VM -> chain but stop at the VM.

I think doing 2. has a few advantages:
1. We do not need to panic in the mock implementation. Both mock and production return errors back to the contract. This makes reviewing the system easier. 
2. I'm not entirely sure that a malformed pubkey or signature is always a programming error. If a contract gets those from external input it might want to be able to handle the error cases in a structured way.

I wannted to avoid going from 1. to 2. by only adding the cost (infrastructure, string processing) without adding the value (possibility for the contract developer to handle errors). I think the enum solution is a great and cheap way to get 2.

Reply:
> I'm glad we're having this discussion. It is an important one and a hard one.

I agree. It's also something that is best NOT left for future discussión, and it affects the (public) API directly.

OK. Will try to do 2. according to your suggestions.
I was just thinking that as it is now it's already an improvement, as the contract can at least return the error message to the user directly (as a `StdError`). So that the user don't have to search for and look it up in the transaction logs.